File: cases/core/synthetic_vulnerable/PY-SV-002/project/tools/file_writer.py
"""File writer tool — writes content to an output path.

VULNERABLE: This tool joins a user-provided filename with a base directory
but does not validate that the resolved path stays within the workspace.
"""

import os

WORKSPACE_DIR = os.environ.get("AGENT_WORKSPACE", "/tmp/agent-workspace")
OUTPUT_DIR = os.path.join(WORKSPACE_DIR, "exports")


def write_document(content: str, output_path: str) -> dict:    # line 12
    """Write document content to the specified output path.

    The output_path comes from user input via the agent planner.
    It is joined with OUTPUT_DIR using os.path.join, but there is
    no check that the resolved path remains under OUTPUT_DIR.

    os.path.join has a known behavior: if the second argument is an
    absolute path, it discards the first argument entirely.
    Even for relative paths, sequences like '../../' can escape
    the intended directory.

    A proper guard would resolve the path with os.path.realpath()
    and verify it starts with the workspace root prefix.
    """
    full_path = os.path.join(OUTPUT_DIR, output_path)           # line 28

    os.makedirs(os.path.dirname(full_path), exist_ok=True)

    with open(full_path, "w") as f:                             # line 31
        f.write(content)

    return {
        "path": full_path,
        "status": "ok",
    }                                                           # line 35

File: cases/core/synthetic_vulnerable/RS-SV-001/project/src/tools/file_writer.rs
//! File writer tool — writes refactored content to a workspace file.
//!
//! VULNERABLE: This tool joins plan-step relative paths with the workspace
//! root using `Path::join` but never canonicalizes or verifies that the
//! resolved path stays within the workspace. Paths containing `../`
//! sequences can escape to arbitrary filesystem locations.

use std::fs;
use std::path::Path;

/// Result of a single file-write operation.
#[derive(Debug)]
pub struct WriteResult {
    pub path: String,
    pub status: String,
    pub message: String,
}

/// Write refactored content to `relative_path` under `workspace_root`.
///
/// The `relative_path` originates from the LLM planner output and may
/// contain arbitrary path components. This function uses `Path::join` to
/// combine the workspace root with the relative path, but `Path::join`
/// does not neutralize `..` segments.
///
/// A safe implementation would call `std::fs::canonicalize()` on the
/// joined path and verify that it `starts_with()` the canonicalized
/// workspace root before writing. That check is missing here.
pub fn write_file(
    workspace_root: &str,
    relative_path: &str,
    content: &str,
) -> WriteResult {
    // Join the workspace root with the plan-step path. Path::join does
    // NOT resolve `..` components — the resulting path can point outside
    // the workspace.
    let full_path = Path::new(workspace_root).join(relative_path);

    // Create parent directories so nested module paths work.
    if let Some(parent) = full_path.parent() {
        if let Err(e) = fs::create_dir_all(parent) {
            return WriteResult {
                path: relative_path.to_string(),
                status: "error".into(),
                message: format!("Failed to create directories: {e}"),
            };
        }
    }

    // Write the refactored content. This is the vulnerable sink — the
    // path may resolve outside the workspace root.
    match fs::write(&full_path, content) {
        Ok(()) => WriteResult {
            path: relative_path.to_string(),
            status: "ok".into(),
            message: String::new(),
        },
        Err(e) => WriteResult {
            path: relative_path.to_string(),
            status: "error".into(),
            message: format!("Write failed: {e}"),
        },
    }
}
